File: Word Search.cpp
class Solution {
public:
    /**
     * @param board: A list of lists of character
     * @param word: A string
     * @return: A boolean
     */
    bool exist(vector<vector<char> > &board, string word) {
        // write your code here
        
        if (board.empty())
        {
            return false;
        }
        vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size(), false));
        for (int i = 0; i < board.size(); ++i)
        {
            for (int j = 0; j < board[0].size(); ++j)
            {
                if (search(i, j, 0, visited, word, board))
                {
                    return true;
                }
            }
        }
        return false;
    }
    bool search(int row, int col, int start, vector<vector<bool>> &visited, const string &word, const vector<vector<char>> &board)
    {
        if (start == word.length())
        {
            return true;
        }
        if (row < 0 || row >= board.size() || col < 0 || col >= board[0].size() || visited[row][col])
        {
            return false;
        }
        
        if (word[start] != board[row][col])
        {
            return false;
        }
        
        visited[row][col] = true;
        vector<pair<int, int>> dicts{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (const auto &p : dicts)
        {
            if (search(row + p.first, col + p.second, start + 1, visited, word, board))
            {
                return true;
            }
        }
        visited[row][col] = false;
        return false;
        
        
        
    }
};

File: Word Search.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        # write your code here
        def dfs(row, col, pos, visited, rows, cols, length, board, word, DICTS):
            if row < 0 or row >= rows or col < 0 or col >= cols or visited[row][col]:
                return False
            if board[row][col] != word[pos]:
                return False
            pos += 1
            if pos == length:
                return True
            visited[row][col] = True
            for x, y in DICTS:
                nx, ny = row + x, col + y
                if dfs(nx, ny, pos, visited, rows, cols, length, board, word, DICTS):
                    return True
            visited[row][col] = False
            return False
            
        
        if not board or not board[0]:
            return False
        rows, cols = len(board), len(board[0])
        length = len(word)
        DICTS = ((-1, 0), (1, 0), (0, -1), (0, 1))
        visited = [[False] * cols for __ in range(rows)]
        for i in range(rows):
            for j in range(cols):
                if dfs(i, j, 0, visited, rows, cols, length, board, word, DICTS):
                    return True
        return False
            